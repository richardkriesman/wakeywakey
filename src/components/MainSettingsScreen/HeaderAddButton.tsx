import React, { ReactNode } from "react";
import { StyleSheet } from "react-native";
import { Button, Icon } from "react-native-elements";

import Colors from "../../constants/Colors";

export interface HeaderAddButtonProps {
    onPress: VoidFunction;
}

export class HeaderAddButton extends React.Component<HeaderAddButtonProps> {

    public constructor(props: HeaderAddButtonProps) {
        super(props);
    }

    public render(): ReactNode {
        return (
            <Button
                type="clear"
                buttonStyle={styles.topRightButton}
                onPress={this.props.onPress.bind(this)}
                icon={<Icon name="add" type="ionicons" color={Colors.appleButtonBlue} />}
            />
        );
    }

}

const styles = StyleSheet.create({
    topRightButton: {
        marginRight: 10
    }
});
