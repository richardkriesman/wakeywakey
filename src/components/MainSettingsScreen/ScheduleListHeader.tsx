import React, { ReactNode } from "react";
import { StyleSheet } from "react-native";
import { ListItem } from "react-native-elements";

export interface ScheduleListHeaderProps {
    title: string;
}

export class ScheduleListHeader extends React.Component<ScheduleListHeaderProps> {

    public constructor(props: ScheduleListHeaderProps) {
        super(props);
    }

    public render(): ReactNode {
        return (
            <ListItem
                containerStyle={styles.container}
                title={this.props.title}
                titleStyle={styles.title}
            />
        );
    }

}

const styles = StyleSheet.create({
    container: {
        borderBottomWidth: StyleSheet.hairlineWidth,
        borderColor: "#e3e3e3"
    },
    title: {
        color: "#5e5e5e",
        fontWeight: "bold"
    }
});
