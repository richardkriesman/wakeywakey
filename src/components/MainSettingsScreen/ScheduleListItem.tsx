import React, { ReactNode } from "react";
import { Icon, ListItem } from "react-native-elements";

export interface ScheduleListItemProps {
    enabled: boolean;
    title: string;

    onSwitchToggled: (enabled: boolean) => void;
}

export interface ScheduleListItemState {
    enabled: boolean;
}

export class ScheduleListItem extends React.Component<ScheduleListItemProps, ScheduleListItemState> {

    public constructor(props: ScheduleListItemProps) {
        super(props);
    }

    public componentWillMount(): void {
        this.setState({ enabled: this.props.enabled });
    }

    public onSwitchValueChanged(isEnabledNow: boolean) {
        this.forceEnabled(isEnabledNow);
        this.props.onSwitchToggled(isEnabledNow);
    }

    public forceEnabled(e: boolean) {
        this.setState({ enabled: e });
    }

    public render(): ReactNode {
        return (
            <ListItem
                title={this.props.title}
                rightIcon={<Icon name="arrow-forward" type="ionicons"/>}
                switch={{
                    onValueChange: this.onSwitchValueChanged.bind(this),
                    value: this.state.enabled
                }}
            />
        );
    }

}
