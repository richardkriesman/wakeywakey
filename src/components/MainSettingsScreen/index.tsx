/**
 * @module components
 */

export { ScheduleListHeader } from "./ScheduleListHeader";
export { ScheduleListItem } from "./ScheduleListItem";
