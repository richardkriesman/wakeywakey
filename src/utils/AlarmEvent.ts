import { Alarm } from "../models/Alarm";

/**
 * The type of {@link AlarmEventType} being fired
 */
export enum AlarmEventType {
    Sleep,
    Wake,
    GetUp
}

export class AlarmEvent {

    public readonly alarm: Alarm;
    public readonly type: AlarmEventType;

    private readonly snoozeFn: (event: AlarmEvent) => Promise<void>;

    public constructor(alarm: Alarm, type: AlarmEventType, snoozeFn: (event: AlarmEvent) => Promise<void>) {
        this.alarm = alarm;
        this.snoozeFn = snoozeFn;
        this.type = type;
    }

    /**
     * Schedules the {@link AlarmEvent} to be replayed after its specified snooze time.
     */
    public snooze(): Promise<void> {
        return this.snoozeFn(this);
    }

}
