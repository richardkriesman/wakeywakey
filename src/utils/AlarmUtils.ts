import { AlarmModel } from "../models/AlarmModel";

export function getAlarmTitle(alarm: AlarmModel): string {
    const start = formatTime(alarm.sleepTime);
    const end = formatTime(alarm.getUpTime);
    return `${start} - ${end}`;
}

export function getAlarmSubtitle(alarm: AlarmModel): string {
    return alarm.days.join(", ");
}

export function formatTime(date: Date): string {
    return `${padTime(date.getHours())}:${padTime(date.getMinutes())}`;
}

export function padTime(time: number): string {
    return ("0" + time).slice(-2);
}

const AlarmUtils = {
    formatTime,
    getAlarmSubtitle,
    getAlarmTitle,
    padTime
};

export default AlarmUtils;
