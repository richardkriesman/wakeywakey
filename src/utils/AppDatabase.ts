/**
 * @module utils
 */

import {FileSystem, SQLite} from "expo";
import * as Log from "./Log";

const DATABASE_NAME: string = "app";
const DATABASE_LOG_TAG: string = "Database";

export class AppDatabase {

    public static get initCount(): number {
        return this._initCount;
    }

    /**
     * Creates and opens a connection to the app's SQLite database, creating any tables that don't exist.
     */
    public static async init(): Promise<AppDatabase> {
        const appDb = new AppDatabase();

        // create preferences table
        if (!(await appDb.doesTableExist("preferences"))) {
            console.debug(`Database table "preferences" does not exist, creating it`);
            await appDb.execute(`
                CREATE TABLE preferences
                (
                    name VARCHAR not null
                        constraint preferences_pk primary key,
                    value VARCHAR not null
                )
            `);
        }

        AppDatabase._initCount++;
        return appDb;
    }

    /**
     * Tracks the number of database initializations. Used to catch accidental re-initializations when the connection
     * is not properly passed between screens.
     */
    private static _initCount: number = 0;

    private db: Database;

    private constructor() {
        this.db = SQLite.openDatabase(DATABASE_NAME);
        Log.info(DATABASE_LOG_TAG, `Opened database at ${FileSystem.documentDirectory}SQLite/${DATABASE_NAME}`);
    }

    /**
     * Determines whether a table with the given name exists in the database.
     *
     * @param name The table name
     */
    public async doesTableExist(name: string): Promise<boolean> {
        const result: SQLResultSet = await this.execute(`
            SELECT *
            FROM sqlite_master
            WHERE
                tbl_name = ?
        `, [name]);
        return result.rows.length > 0;
    }

    /**
     * Returns the value from the key/value table. Returns `null` if no such value exists.
     *
     * Note: This should only be used for storing general preference data that would not be appropriate in a relational
     * schema. Do not use the key/value table for storing models.
     *
     * @param key Unique key identifying the key/value pair.
     */
    public async getPreference(key: string): Promise<string|null> {
        const result: SQLResultSet = await this.execute(`
            SELECT *
            FROM preferences
            WHERE
                name = ?
        `, [key]);
        if (result.rows.length > 0) {
            const value: string = result.rows.item(0).value;
            Log.info(DATABASE_LOG_TAG, `GET ${key}: ${value}`);
            return value;
        } else {
            Log.info(DATABASE_LOG_TAG, `GET ${key}: (null)`);
            return null;
        }
    }

    /**
     * Executes a SQL query against the database, returning a {@link SQLResultSet} containing the resultant rows.
     *
     * @param sql The SQL query to execute.
     * @param args An array of values to substitute for "?" placeholders in the SQL statement.
     */
    public execute(sql: DOMString, args: ObjectArray = []): Promise<SQLResultSet> {
        return new Promise<SQLResultSet>((resolve, reject) => {
            this.db.transaction((transaction) => { // wrap the query in a transaction
                transaction.executeSql(sql, args, (_, resultSet) => { // success, return result set
                    resolve(resultSet);
                }, (_, error) => { // error, reject
                    console.log(error);
                    reject(error);
                    return true;
                });
            });
        });
    }

    /**
     * Sets the value for a key in the key/value table. If a value already exists, it will be overwritten.
     *
     * Note: This should only be used for storing general preference data that would not be appropriate in a relational
     * schema. Do not use the key/value table for storing models.
     *
     * @param key Unique key
     * @param value Value to set for the key
     */
    public async setPreference(key: string, value: string): Promise<void> {
        await this.execute(`
            INSERT OR REPLACE INTO preferences
                (name, value)
            VALUES
                (?, ?);
        `, [key, value]);
        Log.info(DATABASE_LOG_TAG, `SET ${key}: ${value}`);
    }

}
