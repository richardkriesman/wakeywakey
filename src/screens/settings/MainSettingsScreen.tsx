import React, { ReactNode } from "react";
import { SectionList } from "react-native";
import { NavigationScreenProps, StackActions } from "react-navigation";

import { HeaderAddButton } from "../../components/MainSettingsScreen/HeaderAddButton";
import { ScheduleListHeader } from "../../components/MainSettingsScreen/ScheduleListHeader";
import { ScheduleListItem } from "../../components/MainSettingsScreen/ScheduleListItem";
import { TestAlarms } from "../../models/AlarmModel";
import { ScheduleModel } from "../../models/ScheduleModel";
import { HeaderButtonRight } from "../../utils/screen/NavigationOptions";
import { UIScreen } from "../../utils/screen/UIScreen";

/**
 * Main settings screen state. Includes schedule states.
 * @author Shawn Lutch
 */
export interface MainSettingsScreenState {
    schedules: ScheduleModel[];
}

/**
 * Properties of each list item.
 * @author Shawn Lutch
 */
export interface SettingsListItemProps {
    title: string;
}

/**
 * Main Settings screen.
 * @author Shawn Lutch, Miika Raina
 */
@HeaderButtonRight((screen) =>
    <HeaderAddButton onPress={() => screen.present("EditSchedule", { title: "Add Schedule" })} />)
export default class MainSettingsScreen extends UIScreen<{}, MainSettingsScreenState> {

    public constructor(props: NavigationScreenProps) {
        super(props);
    }

    public componentWillMount(): void {
        // TODO properly load schedules from disk
        this.setState({ schedules: testSchedulesList });
    }

    public onScheduleItemToggled(key: number, newEnabled: boolean): void {
        const newSchedules: ScheduleModel[] = this.state.schedules;
        newSchedules[key].enabled = newEnabled;

        if (newEnabled) {
            newSchedules.forEach((s: ScheduleModel) => {
                if (s.key !== key) {
                    s.enabled = false;
                    s.listItemRef.forceEnabled(false);
                }
            });
        }

        this.setState({ schedules: newSchedules });
    }

    public onScheduleItemPressed(key: number): void {
        this.props.navigation.dispatch(StackActions.push({
            params: {
                schedule: this.state.schedules[key],
                title: this.state.schedules[key].name
            },
            routeName: "EditSchedule"
        }));
    }

    public renderContent(): ReactNode {
        return (
            <SectionList
                renderItem={({ item }) => (
                    <ScheduleListItem
                        ref={(me: ScheduleListItem) => {
                            this.state.schedules[item.key].listItemRef = me;
                        }}
                        onPress={this.onScheduleItemPressed.bind(this, item.key)}
                        onSwitchToggled={this.onScheduleItemToggled.bind(this, item.key)}
                        title={item.name}
                        enabled={item.enabled}
                    />
                )}
                renderSectionHeader={({ section }) => <ScheduleListHeader title={section.title}/>}
                sections={[{ data: this.state.schedules, title: "Schedules" }]}
            />
        );
    }
}

// TODO save and load schedules
const testSchedulesList: ScheduleModel[] = [
    {
        alarms: TestAlarms,
        enabled: true,
        key: 0,
        name: "Test Schedule 1"
    },
    {
        alarms: TestAlarms,
        enabled: false,
        key: 1,
        name: "Test Schedule 2"
    }
];
