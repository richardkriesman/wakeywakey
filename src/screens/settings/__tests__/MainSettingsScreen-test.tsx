import * as React from "react";
import Renderer from "react-test-renderer";
import { TestEnvironment } from "../../../utils/testing";
import MainSettingsScreen from "../MainSettingsScreen";

let env: TestEnvironment;
beforeEach((done) => {
    TestEnvironment.init()
        .then((newEnv) => {
            env = newEnv;
            done();
        });
});

it("renders properly", () => {
    const tree = Renderer.create(<MainSettingsScreen navigation={env.navigationProp}/>);
    expect(tree.toJSON()).toMatchSnapshot();
});

describe("toggles", () => {
    it("handles toggling", () => {
        const screen = new MainSettingsScreen({ navigation: env.navigationProp });

        screen.componentWillMount();
        screen.render();

        const setState = jest.spyOn(screen, "setState");
        screen.onScheduleItemToggled(screen.state.schedules.length - 1, true);
        expect(setState).toHaveBeenCalledTimes(1);
    });
});
