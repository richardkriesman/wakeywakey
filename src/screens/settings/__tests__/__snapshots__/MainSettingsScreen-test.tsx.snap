// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Modal
      animationType="none"
      hardwareAccelerated={false}
      hideModalContentWhileAnimating={false}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onRequestClose={[Function]}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeThreshold={100}
      transparent={true}
      visible={false}
    >
      <View
        accessible={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "black",
            "bottom": 0,
            "height": 1334,
            "left": 0,
            "opacity": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 750,
          }
        }
      />
      <View
        onLayout={[Function]}
        pointerEvents="box-none"
        style={
          Array [
            Array [
              Object {
                "margin": 37.5,
                "transform": Array [
                  Object {
                    "translateY": 0,
                  },
                ],
              },
              Object {
                "flex": 1,
                "justifyContent": "center",
              },
              undefined,
              Object {
                "margin": 0,
              },
            ],
            Object {
              "paddingBottom": 0,
            },
          ]
        }
      >
        <View
          hideModalContentWhileAnimating={false}
          onModalHide={[Function]}
          onModalWillHide={[Function]}
          onModalWillShow={[Function]}
          pointerEvents="box-none"
          scrollOffset={0}
          scrollOffsetMax={0}
          scrollTo={null}
          style={
            Object {
              "flex": 1,
              "justifyContent": "center",
              "margin": 37.5,
              "transform": Array [
                Object {
                  "translateY": 0,
                },
              ],
            }
          }
          supportedOrientations={
            Array [
              "portrait",
              "landscape",
            ]
          }
          swipeThreshold={100}
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#ffffff",
                  "borderRadius": 4,
                  "minWidth": 270,
                }
              }
            >
              <View
                style={
                  Object {
                    "padding": 24,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontSize": 17,
                      "fontWeight": "600",
                      "marginBottom": 8,
                    }
                  }
                >
                  Create schedule
                </Text>
                <Text
                  style={
                    Object {
                      "fontSize": 17,
                      "marginBottom": 8,
                    }
                  }
                >
                  Type a name for the new schedule:
                </Text>
                <View
                  style={
                    Object {
                      "paddingHorizontal": 0,
                      "width": "100%",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "borderBottomColor": "#d9d9d9",
                        "borderBottomWidth": 1,
                        "borderColor": "#86939e",
                        "flexDirection": "row",
                        "transform": Array [
                          Object {
                            "translateX": 0,
                          },
                        ],
                      }
                    }
                  >
                    <TextInput
                      allowFontScaling={true}
                      maxLength={30}
                      onChangeText={[Function]}
                      shake={true}
                      style={
                        Object {
                          "alignSelf": "center",
                          "color": "black",
                          "flex": 1,
                          "fontSize": 18,
                          "minHeight": 40,
                        }
                      }
                      testID="RNE__Input__text-input"
                      underlineColorAndroid="transparent"
                      value=""
                    />
                  </View>
                </View>
              </View>
              <View
                style={
                  Object {
                    "borderTopColor": "#d9d9d9",
                    "borderTopWidth": 1,
                    "flexDirection": "row",
                    "justifyContent": "flex-end",
                    "paddingBottom": 4,
                    "paddingLeft": 18,
                    "paddingRight": 18,
                    "paddingTop": 4,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "borderRadius": 3,
                    }
                  }
                >
                  <View
                    accessible={true}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "transparent",
                          "borderColor": "#2089dc",
                          "borderRadius": 3,
                          "borderWidth": 0,
                          "flexDirection": "row",
                          "justifyContent": "center",
                          "padding": 8,
                        }
                      }
                    >
                      <Text
                        style={
                          Object {
                            "color": "#2089dc",
                            "fontSize": 18,
                            "paddingBottom": 1,
                            "paddingTop": 2,
                            "textAlign": "center",
                          }
                        }
                      >
                        Cancel
                      </Text>
                    </View>
                  </View>
                </View>
                <View
                  style={
                    Object {
                      "borderRadius": 3,
                    }
                  }
                >
                  <View
                    accessible={true}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "transparent",
                          "borderColor": "#2089dc",
                          "borderRadius": 3,
                          "borderWidth": 0,
                          "flexDirection": "row",
                          "justifyContent": "center",
                          "padding": 8,
                        }
                      }
                    >
                      <Text
                        style={
                          Object {
                            "color": "#2089dc",
                            "fontSize": 18,
                            "paddingBottom": 1,
                            "paddingTop": 2,
                            "textAlign": "center",
                          }
                        }
                      >
                        OK
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </Modal>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text />
      <Text
        style={
          Object {
            "fontSize": 17,
          }
        }
      >
        No schedules yet
      </Text>
      <Text
        style={
          Object {
            "color": "#5e5e5e",
            "fontSize": 15,
          }
        }
      >
        Create a schedule to set alarms
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
